### PR TITLE
Allow customizing entry mode icons in TimePicker

### LINKED ISSUE #156942: TimePicker: Allow customizing entry mode icons
### Steps to reproduce

1. showTimePicker(context: context)
2. see no options for changing icons
3. `icon: Icon(_entryMode.value == TimePickerEntryMode.dial ? Icons.keyboard_outlined : Icons.access_time),`

### Expected results

Able to change icons, like in showDatePicker()

### Actual results

Cannot change icons.

### Code sample

<details open><summary>Code sample</summary>

```dart
await showTimePicker(context: context)
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.3, on macOS 14.6.1 23G93 darwin-arm64, locale en-CZ)
    • Flutter version 3.24.3 on channel stable at /opt/homebrew/Caskroom/flutter/3.19.5/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (5 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/michalnemec/Library/Android/sdk
    • Platform android-34, build-tools 33.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.98.0


[✓] Connected device (6 available)
    • Mi A2 (mobile)                  • 1f95281c                  • android-arm64  • Android 10 (API 29)
    • iPhone (mobile)                 • 00008101-000B044A3C02001E • ios            • iOS 18.0.1 22A3370
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.6.1 23G93 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 120.0.6099.216

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by syedaniq**
This PR adds two new properties to the Material Design `TimePicker`: `switchToInputEntryModeIcon` and `switchToTimerEntryModeIcon`. These allow developers to customize the icons for toggling between dial and input modes, with default icons (`Icons.keyboard_outlined` and `Icons.access_time`) used if none are provided.

Fixes #156942 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/time_picker.dart [hunk 1]
```diff
@@ -2219,6 +2221,12 @@ class TimePickerDialog extends StatefulWidget {
   /// Callback called when the selected entry mode is changed.
   final EntryModeChangeCallback? onEntryModeChanged;
 
+  /// {@macro flutter.material.time_picker.switchToInputEntryModeIcon}
+  final Icon? switchToInputEntryModeIcon;
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/time_picker.dart:2225]: Since the default is a const icon, can we instead make these non-null and provide them as default arguments in the constructor?

I like that we provide an `Icon` widget; users can then tweak the icon, size, and color, too.

File: packages/flutter/lib/src/material/time_picker.dart [hunk 2]
```diff
@@ -2540,6 +2552,8 @@ class _TimePicker extends StatefulWidget {
     this.entryMode = TimePickerEntryMode.dial,
     this.orientation,
     this.onEntryModeChanged,
+    this.switchToInputEntryModeIcon,
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/time_picker.dart:2555]: Same idea as above

File: packages/flutter/lib/src/material/time_picker.dart [hunk 3]
```diff
@@ -2941,6 +2961,22 @@ class _TimePickerState extends State<_TimePicker> with RestorationMixin {
 /// parameter to override the default and force the dialog to appear in either
 /// portrait or landscape mode.
 ///
+/// {@template flutter.material.time_picker.switchToInputEntryModeIcon}
+/// An optional [switchToInputEntryModeIcon] argument can be used to
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/time_picker.dart:2965]:
```suggestion
/// The [switchToInputEntryModeIcon] argument can be used to customize
/// the input method icon that is shown when the [TimePickerEntryMode] is [TimePickerEntryMode.dial].
///
/// Defaults to an [Icon] widget with [Icons.keyboard_outlined] as icon.
```

File: packages/flutter/lib/src/material/time_picker.dart [hunk 4]
```diff
@@ -2941,6 +2961,22 @@ class _TimePickerState extends State<_TimePicker> with RestorationMixin {
...
+/// display a custom Icon in the corner of the dialog
+/// when [TimePickerEntryMode] is [TimePickerEntryMode.dial]. Clicking on
+/// icon changes the [TimePickerEntryMode] to [TimePickerEntryMode.input].
+/// If null, `Icon(keyboard_outlined)` is used.
+/// {@endtemplate}
+///
+/// {@template flutter.material.time_picker.switchToTimerEntryModeIcon}
+/// An optional [switchToTimerEntryModeIcon] argument can be used to
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/time_picker.dart:2973]:
```suggestion
/// The [switchToTimerEntryModeIcon] argument can be used to customize
/// the input method icon that is shown when the [TimePickerEntryMode] is [TimePickerEntryMode.input].
///
/// Defaults to an [Icon] widget with [Icons.access_time] as icon.
```

File: packages/flutter/lib/src/material/time_picker.dart [hunk 5]
```diff
@@ -3015,6 +3051,8 @@ Future<TimeOfDay?> showTimePicker({
   EntryModeChangeCallback? onEntryModeChanged,
   Offset? anchorPoint,
   Orientation? orientation,
+  final Icon? switchToInputEntryModeIcon,
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/time_picker.dart:3054]: Same idea as above for the non-null defaults.

Also, was making these `final` a mistake? The others are not final either, and we do not enforce that arguments in constructors are final, so I think that the `final` keywords can be removed here?

File: packages/flutter/test/material/time_picker_test.dart [hunk 6]
```diff
@@ -1646,6 +1646,45 @@ void main() {
         expect(find.text(errorInvalidText), findsOneWidget);
       });
 
+      testWidgets('TimePicker shows default icons when no custom icons are provided', (WidgetTester tester) async {
+        await tester.pumpWidget(MaterialApp(
```
Comment (navaronbracke) [inline packages/flutter/test/material/time_picker_test.dart:1650]: This test will need to be reformatted. The indentation looks off.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (syedaniq) [reply packages/flutter/lib/src/material/time_picker.dart]: Hi, thanks for your time. The approach you're suggesting makes much more sense. I initially coded it that way because it was done similarly in `date_picker.dart`
https://github.com/flutter/flutter/blob/0500af2b5f750925e355266360307ff4e69c0954/packages/flutter/lib/src/material/date_picker.dart#L441-L442

File: packages/flutter/lib/src/material/time_picker.dart (hunk 5 above)
Comment (syedaniq) [reply packages/flutter/lib/src/material/time_picker.dart]: I was again inspired by the approach done in `date_picker.dart`, maybe we can also use default arguments there in a new issue 
https://github.com/flutter/flutter/blob/0500af2b5f750925e355266360307ff4e69c0954/packages/flutter/lib/src/material/date_picker.dart#L1152-L1155

**Review by navaronbracke — COMMENTED** [review]
This looks good to me. The failing CI task seems unrelated to your PR, though.

Seems like it is also failing for other PR's.

**Comment by syedaniq** [comment]
Thanks for your time and the helpful suggestions @navaronbracke, really appreciate it.
The checks were all sucessfull this time 🥳

File: packages/flutter/lib/src/material/time_picker.dart [hunk 7]
```diff
@@ -2219,6 +2221,12 @@ class TimePickerDialog extends StatefulWidget {
   /// Callback called when the selected entry mode is changed.
   final EntryModeChangeCallback? onEntryModeChanged;
 
+  /// {@macro flutter.material.time_picker.switchToInputEntryModeIcon}
+  final Icon switchToInputEntryModeIcon;
+
+  /// {@macro flutter.material.time_picker.switchToTimerEntryModeIcon}
+  final Icon switchToTimerEntryModeIcon;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/time_picker.dart:2228]: I was going to leave a suggestion here, but since this is set up to mirror the date picker, it's probably best to leave as-is for consistency.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/time_picker.dart]: +1 to having it be non-nullable, cause why not?

Thanks for changing it here!

**Review by nate-thegrate — APPROVED** [review]
LGTM

Thanks for the contribution!

File: packages/flutter/lib/src/material/time_picker.dart (hunk 7 above)
Comment (TahaTesser) [inline packages/flutter/lib/src/material/time_picker.dart:2228]: Could we make these properties nullable? Revert this change to restore the defaults and just override them here

image
 



This way we don't have to provide multiple parameter with defaults values 

image

image

image

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/time_picker.dart]: I would suggest nullable props so we can define defaults once instead of 3 times at three difference places.

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/time_picker.dart]: That's a good point—a single source of truth is easier to maintain, plus being nullable would make it parallel with date_picker.dart 👍

File: packages/flutter/test/material/time_picker_test.dart [hunk 8]
```diff
@@ -1646,6 +1646,45 @@ void main() {
         expect(find.text(errorInvalidText), findsOneWidget);
       });
 
+      testWidgets('TimePicker shows default icons when no custom icons are provided', (WidgetTester tester) async {
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1649]:
```suggestion
      testWidgets('TimePicker default entry icons', (WidgetTester tester) async {
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 9]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+        expect(find.byIcon(Icons.access_time), findsNothing);
+        // Tap the icon to switch to input mode
+        await tester.tap(find.byIcon(Icons.keyboard_outlined));
+        await tester.pumpAndSettle();
+        // Check that the icon for the input mode is displayed
+        expect(find.byIcon(Icons.access_time), findsOneWidget);
+        expect(find.byIcon(Icons.keyboard_outlined), findsNothing);
+      });
+
+      testWidgets('TimePicker uses custom icons if provided', (WidgetTester tester) async {
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1665]:
```suggestion
      testWidgets('Can override TimePicker entry icons', (WidgetTester tester) async {
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 10]
```diff
@@ -1646,6 +1646,45 @@ void main() {
         expect(find.text(errorInvalidText), findsOneWidget);
       });
 
+      testWidgets('TimePicker shows default icons when no custom icons are provided', (WidgetTester tester) async {
+        await tester.pumpWidget(MaterialApp(
+          home: TimePickerDialog(initialTime: TimeOfDay.now()),
+        ));
+
+        // Check that the default icon for the dial mode is displayed
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1654]:
```suggestion
        // Check that the default icon for the dial mode is displayed.
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 11]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+        expect(find.byIcon(Icons.keyboard_outlined), findsOneWidget);
+        expect(find.byIcon(Icons.access_time), findsNothing);
+        // Tap the icon to switch to input mode
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1657]:
```suggestion
        // Tap the icon to switch to input mode.
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 12]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+        await tester.tap(find.byIcon(Icons.keyboard_outlined));
+        await tester.pumpAndSettle();
+        // Check that the icon for the input mode is displayed
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1660]:
```suggestion
        // Check that the icon for the input mode is displayed.
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 13]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+
+        await tester.pumpWidget(MaterialApp(
+          home: TimePickerDialog(
+            initialTime: TimeOfDay.now(),
+            switchToInputEntryModeIcon: customInputIcon,
+            switchToTimerEntryModeIcon: customTimerIcon,
+          ),
+        ));
+
+        // Check that the custom icons are displayed
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1677]:
```suggestion
        // Check that the custom icons are displayed.
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 14]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+        expect(find.byIcon(Icons.text_fields), findsOneWidget);
+        expect(find.byIcon(Icons.watch), findsNothing);
+        // Tap the custom icon to switch to input mode
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1680]:
```suggestion
        // Tap the custom icon to switch to input mode.
```

File: packages/flutter/test/material/time_picker_test.dart [hunk 15]
```diff
@@ -1646,6 +1646,45 @@ void main() {
...
+        await tester.tap(find.byIcon(Icons.text_fields));
+        await tester.pumpAndSettle();
+        // Check that the custom icon for the input mode is displayed
```
Comment (TahaTesser) [inline packages/flutter/test/material/time_picker_test.dart:1683]:
```suggestion
        // Check that the custom icon for the input mode is displayed.
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Almost there!

**Review by TahaTesser — APPROVED** [review]
LGTM

**Comment by MichalNemec** [comment]
when this will get on stable? waiting for it since 3.19.5 (March 28, 2024)

**Comment by nate-thegrate** [comment]
It looks like [the issue](https://github.com/flutter/flutter/issues/156942) was filed on October 16 and then this contribution was merged on October 29.
The current stable branch contains all contributions prior to October 9, so this change missed the cutoff by a few weeks.

Feel free to check out https://docs.flutter.dev/release/archive to get an idea of how often stable releases generally take place. The next one will likely be in about 3 months from now.
